crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -128,37 +129,113 @@ impl<S: States> NextState<S> {
-/// - Runs the [`OnExit(exited_state)`] schedule, if it exists.
-/// - Runs the [`OnTransition { from: exited_state, to: entered_state }`](OnTransition), if it exists.
-/// - Runs the [`OnEnter(entered_state)`] schedule, if it exists.
+/// Label to order all [`OnEnter`], [`OnTransition`], and [`OnExit`] systems.
+///
+/// All `OnEnter` will run before all `OnTransition`,
+/// and all `OnTransition` will run before all `OnExit`.
+///
+/// The order of `OnEnter` systems relative to each other (and so on) is ambiguous by default.
+#[derive(SystemSet, Copy, Clone, Debug, Eq, PartialEq, Hash)]
+pub enum StateTransitionSet {
+    /// Change all the states before invoking transition systems here.
+    BeforeTransition,
+    /// All [`OnEnter`] schedules are executed in this set.
+    OnEnter,
+    /// All [`OnTransition`] schedules are executed in this set.
+    OnTransition,
+    /// All [`OnExit`] schedules are executed in this set.
+    OnExit,
+    /// Cleanup `PrevState` after all transition systems have run.
+    AfterTransition,
+}
+
+/// Run all state transition systems for the given state type.
+///
+/// This function is deprecated. It is not used internally by Bevy,
+/// and will be removed in a future version.
+#[deprecated(since = "0.13.0")]
 pub fn apply_state_transition<S: States>(world: &mut World) {
-    // We want to take the `NextState` resource,
-    // but only mark it as changed if it wasn't empty.
+    apply_state_transition_before_transition::<S>(world);
+    apply_state_transition_on_enter::<S>(world);
+    apply_state_transition_on_transition::<S>(world);
+    apply_state_transition_on_exit::<S>(world);
+    apply_state_transition_after_transition::<S>(world);
+}
+
+/// Change the state before invoking transition systems.
+pub fn apply_state_transition_before_transition<S: States>(world: &mut World) {
+    // Reset the previous state, so the following systems which rely only on `PrevState` won't run.
+    let mut prev_state_resource = world.resource_mut::<PrevState<S>>();
+    if prev_state_resource.bypass_change_detection().0.is_some() {
+        prev_state_resource.0 = None;
+    }
+
     let mut next_state_resource = world.resource_mut::<NextState<S>>();
     if let Some(entered) = next_state_resource.bypass_change_detection().0.take() {
         next_state_resource.set_changed();
 
         let mut state_resource = world.resource_mut::<State<S>>();
         if *state_resource != entered {
             let exited = mem::replace(&mut state_resource.0, entered.clone());
-            // Try to run the schedules if they exist.
-            world.try_run_schedule(OnExit(exited.clone())).ok();
-            world
-                .try_run_schedule(OnTransition {
-                    from: exited,
-                    to: entered.clone(),
-                })
-                .ok();
-            world.try_run_schedule(OnEnter(entered)).ok();
+            world.resource_mut::<PrevState<S>>().0 = Some(exited.clone());
```

Comment:
```suggestion
            world.resource_mut::<PrevState<S>>().0 = Some(exited);
```

I think this clone is unnecessary.